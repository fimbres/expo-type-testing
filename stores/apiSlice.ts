import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

import { IProduct } from './types';

export const apiSlice = createApi({
    reducerPath: 'api',
    tagTypes: ['Product'],
    baseQuery: fetchBaseQuery({
        baseUrl: 'https://6222994f666291106a29f999.mockapi.io/api/v1'
    }),
    endpoints: builder => ({
        getProducts: builder.query({
            query: () => '/products',
            transformResponse: loadedProducts => {
                return loadedProducts as IProduct[];
            }
        })
    })
});

export const { useGetProductsQuery } = apiSlice;
