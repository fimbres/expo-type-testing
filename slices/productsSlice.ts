import { createSelector } from "@reduxjs/toolkit";
import moment from "moment";

import { apiSlice } from "./apiSlice";
import { IProduct } from "../types";

const initialState: IProduct[] = [];

const selectUsersResult = apiSlice.endpoints.getProducts.select({});

const sortProducts = (first: IProduct, second: IProduct) => {
  return moment(second.createdAt).valueOf() - moment(first.createdAt).valueOf();
};

export const selectAllProducts = createSelector(
  selectUsersResult,
  (productsResult) =>
    productsResult.data
      ? [...productsResult.data].sort(sortProducts)
      : initialState
);

export const selectTotalPoints = createSelector(selectAllProducts, (products) =>
  products.reduce((sum, product) => {
    if (product.is_redemption) {
      return (sum -= product.points);
    } else {
      return (sum += product.points);
    }
  }, 0)
);

export const selectProductById = createSelector(
  selectAllProducts,
  (_: IProduct[], productId: number) => productId,
  (products, productId) => products.find((product) => product.id == productId)
);

export const selectRedemptionProducts = createSelector(
  selectAllProducts,
  (products) => products.filter((product) => product.is_redemption)
);

export const selectNotRedemptionProducts = createSelector(
  selectAllProducts,
  (products) => products.filter((product) => !product.is_redemption)
);
