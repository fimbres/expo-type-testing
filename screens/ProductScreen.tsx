import React, { FC } from 'react'
import { Image, ScrollView, StyleSheet, Text, View } from 'react-native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import Numeral from "numeral";
import moment from 'moment';
import 'moment/locale/es';

import { Colors, Styles } from '../constants'
import { RootStackParamList } from '../navigation/types'
import { AppButton, SafeAreaContainer, ScreenHeader } from '../components'
import { useSelector } from 'react-redux'
import { selectProductById } from '../stores/productsSlice'
import { IProduct } from '../types'

const ProductScreen: FC<NativeStackScreenProps<RootStackParamList, "Product">> = ({ navigation, route }) => {
    const { productId } = route.params;
    const product = useSelector<any, IProduct | undefined>(state => selectProductById(state, productId));
    const { image, product: name, points, createdAt } = product!;

    if(!product){
        navigation.navigate("NotFound");
    }

    const getDateLabel = () => {
        const momentDate = moment(createdAt);
        return `Comprado el ${momentDate.format("d")} de ${momentDate.format("MMMM, YYYY")}`
    }

    return (
        <>
            <SafeAreaContainer style={styles.statusBar} />
            <SafeAreaContainer style={{ paddingTop: 0 }}>
            <>
                <ScreenHeader title={name} style={styles.header} />
                <ScrollView style={styles.container}>
                    <View style={styles.imageContainer}>
                        <Image source={{ uri: image }} style={styles.image} />
                    </View>
                    <Text style={[styles.subtitle, { marginTop: 32, marginBottom: 19}]}>Detalles del producto:</Text>
                    <Text style={Styles.textTitleRegular}>{getDateLabel()}</Text>
                    <Text style={[styles.subtitle, { marginTop: 20, marginBottom: 32}]}>Con esta compra acumulaste:</Text>
                    <Text style={Styles.textTitleExtraLarge}>{Numeral(points).format("0,0")} puntos</Text>
                </ScrollView>
                <View style={styles.buttonContainer}>
                    <AppButton title='Aceptar' size='lg' onPress={() => navigation.goBack()} />
                </View>
            </>
            </SafeAreaContainer>
        </>
    )
}

export default ProductScreen;

const styles = StyleSheet.create({
    statusBar: {
        flex: 0,
        backgroundColor: Colors.violet,
    },
    header: {
        paddingTop: 66,
        paddingBottom: 24,
        backgroundColor: Colors.violet,
    },
    container: {
        flex: 1,
        paddingHorizontal: 20,
    },
    imageContainer: {
        width: '100%',
        justifyContent: 'center',
        alignItems: 'center',
        aspectRatio: 1,
        borderRadius: 10,
        marginTop: 19,
        backgroundColor: Colors.white,
        shadowColor: Colors.black,
        shadowOffset: {
            width: 2,
            height: 6,
        },
        shadowOpacity: 0.2,
        shadowRadius: 8,
        elevation: 3,
    },
    image: {
        width: 200,
        height: 200,
    },
    subtitle: {
        ...Styles.textTitleSmall,
        color: Colors.gray,
    },
    buttonContainer: {
        flexDirection: 'row',
        marginBottom: 40,
        paddingHorizontal: 20,
    }
});
