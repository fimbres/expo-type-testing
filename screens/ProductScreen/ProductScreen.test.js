/**
 * @jest-environment jsdom
 */
import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react-native";
import { ApiProvider } from "@reduxjs/toolkit/dist/query/react";

import ProductScreen from ".";
import { apiSlice } from "../../slices/apiSlice";

const product = {
  createdAt: "2022-12-08T18:54:56.243Z",
  id: 5,
  image: "https://loremflickr.com/640/480/people",
  is_redemption: true,
  points: 69814,
  product: "Rustic Rubber Bacon",
};

describe("<ProductScreen />", () => {
  it("render text correctly", () => {
    render(
      <ApiProvider api={apiSlice}>
        <ProductScreen route={{ params: { product } }} />
      </ApiProvider>
    );

    const firstText = screen.getByText("Detalles del producto:");
    const secondText = screen.getByText("Con esta compra acumulaste:");
    const thirdText = screen.getByText("Aceptar");

    expect(firstText).toBeDefined();
    expect(secondText).toBeDefined();
    expect(thirdText).toBeDefined();
  });

  it("press button", () => {
    const navigation = {
      goBack: () => {},
    };

    spyOn(navigation, "goBack");

    const { getByTestId } = render(
      <ApiProvider api={apiSlice}>
        <ProductScreen
          navigation={navigation}
          route={{ params: { product } }}
        />
      </ApiProvider>
    );

    const button = getByTestId("app-button");

    fireEvent.press(button);

    expect(navigation.goBack).toHaveBeenCalled();
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(
        <ApiProvider api={apiSlice}>
          <ProductScreen route={{ params: { product } }} />
        </ApiProvider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
