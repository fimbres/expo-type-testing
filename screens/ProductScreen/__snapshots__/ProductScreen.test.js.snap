// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductScreen /> renders correctly 1`] = `
Array [
  <RCTSafeAreaView
    style={
      Array [
        Object {
          "backgroundColor": "#F8F8F8",
          "flex": 1,
          "paddingTop": 0,
        },
        Object {
          "backgroundColor": "#CFD6FF",
          "flex": 0,
        },
      ]
    }
  />,
  <RCTSafeAreaView
    style={
      Array [
        Object {
          "backgroundColor": "#F8F8F8",
          "flex": 1,
          "paddingTop": 0,
        },
        Object {
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "paddingHorizontal": 20,
            "paddingTop": 24,
          },
          Object {
            "backgroundColor": "#CFD6FF",
            "paddingBottom": 24,
            "paddingTop": 66,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#020202",
            "fontFamily": "avenir-black",
            "fontSize": 24,
            "fontWeight": "800",
            "lineHeight": 24,
          }
        }
      >
        Rustic Rubber Bacon
      </Text>
    </View>
    <RCTScrollView
      style={
        Object {
          "flex": 1,
          "paddingHorizontal": 20,
        }
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "aspectRatio": 1,
              "backgroundColor": "#FFFFFF",
              "borderRadius": 10,
              "elevation": 3,
              "justifyContent": "center",
              "marginTop": 19,
              "shadowColor": "#020202",
              "shadowOffset": Object {
                "height": 6,
                "width": 2,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 8,
              "width": "100%",
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://loremflickr.com/640/480/people",
              }
            }
            style={
              Object {
                "height": 200,
                "width": 200,
              }
            }
            testID="product-image"
          />
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#9B9898",
                "fontFamily": "avenir-black",
                "fontSize": 14,
                "fontWeight": "800",
                "lineHeight": 19.12,
              },
              Object {
                "marginBottom": 19,
                "marginTop": 32,
              },
            ]
          }
        >
          Detalles del producto:
        </Text>
        <Text
          style={
            Object {
              "color": "#020202",
              "fontFamily": "avenir-black",
              "fontSize": 16,
              "fontWeight": "800",
              "lineHeight": 21.86,
            }
          }
        >
          Comprado el 4 de diciembre, 2022
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#9B9898",
                "fontFamily": "avenir-black",
                "fontSize": 14,
                "fontWeight": "800",
                "lineHeight": 19.12,
              },
              Object {
                "marginBottom": 32,
                "marginTop": 20,
              },
            ]
          }
        >
          Con esta compra acumulaste:
        </Text>
        <Text
          style={
            Object {
              "color": "#020202",
              "fontFamily": "avenir-black",
              "fontSize": 24,
              "fontWeight": "800",
              "lineHeight": 24,
            }
          }
        >
          69,814 puntos
        </Text>
      </View>
    </RCTScrollView>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
            "marginVertical": 40,
          },
          Object {
            "paddingHorizontal": 20,
          },
        ]
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "#334FFA",
              "borderRadius": 10,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "paddingVertical": 14,
            },
            false,
          ]
        }
        testID="app-button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
                "fontFamily": "avenir-black",
                "fontSize": 16,
                "fontWeight": "800",
                "lineHeight": 21.86,
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
        >
          Aceptar
        </Text>
      </View>
    </View>
  </RCTSafeAreaView>,
]
`;
